utils/providerAPI/auth.ts

```diff
@@ -0,0 +1,83 @@
+import axios from "axios";
+import {getAuthInfoFromDb, saveAuthInfoToDb} from '../db/authInfo';
+import AuthInfo from "../../types/AuthInfo";
+
+
+export const bitbucketAccessToken = async function(providerAccountId: string, userId: string): Promise<string | null> {
+	console.info("Providing access token...", providerAccountId, userId);
+    const provider = "bitbucket";
+	// get authinfo from db
+	const authInfo: AuthInfo | null = await getAuthInfoFromDb(userId).catch(_ => {
+        console.error("Failed to fetch auth info from the database.");
+        return null;
+    });
+	if (!authInfo || !authInfo[provider] || !authInfo[provider][providerAccountId]) {
+        console.error("Invalid or empty auth info: ", authInfo);
+        return null;
+    }
+	const account = authInfo[provider][providerAccountId];
+	// Check if token has expired.
+    const currentTime = Math.floor(Date.now() / 1000); // Current time in seconds.
+    if (account.expires_at && account.expires_at > currentTime) {
+        console.info("access token not expired yet ",
+            currentTime, account.expires_at);
+        return account.access_token!; // Token is still valid.
+    }
+	// If it has expired, use the refresh token to obtain a new access token.
+	if (!account.refresh_token) {
+        console.error("No refresh token available.");
+        return null;
+    }
+	const newAuthInfo = await bitbucketRefreshToken(account.refresh_token, authInfo, providerAccountId);
+	if (!newAuthInfo || !newAuthInfo[provider] || !newAuthInfo[provider][providerAccountId]) {
+		console.error("No new access token returned from refresh token api", JSON.stringify(newAuthInfo!));
+		return null;
+	}
+	// save new auth info to db
+	await saveAuthInfoToDb(userId, newAuthInfo);
```

Comment:
When we update the auth-info in the database, we should also update the auth info in the session object provided by next-auth because that is used everywhere else (including the browser extension).

High LOGAF